### PR TITLE
Integrate OpusCleaner

### LINKED ISSUE #160: Integrate OpusCleaner
[OpusCleaner](https://github.com/hplt-project/OpusCleaner) is a new data-cleaning tool developed as a part of the [HPLT](https://hplt-project.org/) project.

This tool can help us customize cleaning for different languages and datasets, provide a convenient UI and a way to store cleaning configs. Also, we'll be in sync with the latest developments in the HPLT.

I assume we can implement the following workflow: 
1. Use the UI to develop the cleaning configs
2. Host configs in this repo or in some other place to share with HPLT project
3. Keep our current dataset downloading procedures
4. Use the [standalone cleaning script](https://github.com/hplt-project/OpusCleaner/blob/main/opuscleaner/clean.py) provided by OpusCleaner in the cleaning step of the pipeline instead of the current scripts

**Questions:**
1. Is OpusCleaner stable enough to start integration? Is it already being actively used? It looks like it's under active development
2. Creating a custom cleaning config for each language and dataset is challenging, especially if you don't know the language. Should we have a hierarchy of default configs, for example, defaults for all, defaults for a language, and defaults for a dataset? 
3. Where to store the configs and what are the plans of the developers of HPLT for this? At some point we might want to develop guidelines on how to use the tool and accept community contributions.

### PR DESCRIPTION
**Pull request by eu9ene**
The first pass on integrating [OpusCleaner](https://github.com/hplt-project/OpusCleaner) that's meant to replace our cleaning scripts. 

Included in this PR:
- cleaning parallel corpus
- default config that is similar to our current config
- integration with Snakemake
- integration with TaskCluster
- run in CI
- tested with a full pipeline production config and a default filter on en-ru and Snakemake (all completed except CCMatrix)

Out of scope:
- cleaning monolingual data
- reviewing deeper and improving the default config
- creating the custom filters for a language pair using OpusCleaner UI 
- running experiments to see how the new filtering and custom filters affect translation quality
- migrating the repository of configs to a different place (we might not want to include all the custom configs in the TC artifacts and it seems `resources` doesn't support directories or wildcards)

fixes #160 

Discovered unfixed issues:
 - https://github.com/hplt-project/OpusCleaner/issues/116
 - TaskCluster doesn't redirect stderr to stdout like we do in snakemake, so OpusCleaner logs are not visible by default. I had to redirect manually to a file which is ok but makes it less visible when running manually. We should probably think of a more generic solution because some tools that work with stdin/stdout write log to stderr.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -116,6 +116,19 @@ tensorboard:
 	tensorboard --logdir=$(MODELS) --host=0.0.0.0 &
 	python utils/tb_log_parser.py --prefix=
 
+
+install-opuscleaner:
```
Comment (gregtatum) [inline Makefile:120]: Suggestion(docs): Since the Makefile is a developer-facing entry point into the code, it would be nice to provide a short high-level piece of documentation describing what each these actions are. It would also be nice to mention what OpusCleaner is for someone onboarding to the codebase.

> OpusCleaner is a data cleaner for training data. https://github.com/hplt-project/OpusCleaner

File: pipeline/clean/opuscleaner/configs/default.filters.json [hunk 2]
```diff
@@ -0,0 +1,91 @@
+{
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/configs/default.filters.json:1]: Thought: It's a shame this is JSON as we can't add comments about the configuration and why certain defaults were chosen.

File: pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Starting text in a lot of rows in OpenSubtitles, not everytime in both source and targets
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt:1]: Suggestion: This first comment doesn't make sense to me. It would be nice to re-phrase it using a more complete sentence.

The same goes for "Text in Czech Bible translations that is not in the source"

Comment (gregtatum) [inline pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt:1]: Question: What is this file exactly?

File: pipeline/clean/opuscleaner/generate_filters.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+"""
+Generates filter config for a dataset based on defaults to use in OpusCleaner
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/generate_filters.py:2]: Praise: Thanks for the docs!

File: pipeline/clean/opuscleaner/generate_filters.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+"""
+
+import argparse
+import json
+import os
+from typing import Dict, Optional
+
+
+def find_custom_filter(src: str, trg: str, dataset: str) -> Optional[Dict]:
+    # TODO: we'll likely need to move to a separate repo for those
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/generate_filters.py:12]: Suggestion: If we have follow-ups, it would be good to file them and link the issue here.

File: pyproject.toml [hunk 6]
```diff
@@ -7,6 +7,7 @@ readme = "README.md"
 
 [tool.poetry.dependencies]
 python = "^3.10"
+pip-tools = "^7.3.0"
```
Comment (gregtatum) [inline pyproject.toml:10]: Thought: We should probably figure out the long-term maintenance strategy for our python dependencies, and how we want to manage them.

File: taskcluster/ci/clean-corpus/kind.yml [hunk 7]
```diff
@@ -16,19 +17,33 @@ kind-dependencies:
...
                     - pipeline/clean/clean-corpus.sh
                     - pipeline/clean/tools/deescape-special-chars.perl
                     - pipeline/clean/tools/remove-non-printing-char.perl
                     - pipeline/clean/tools/clean_parallel.py
                     - pipeline/clean/tools/langid_fasttext.py
+                    - pipeline/clean/opuscleaner/generate_filters.py
+                    - pipeline/clean/opuscleaner/clean-corpus.sh
+                    - pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt
+                    - pipeline/clean/opuscleaner/configs/default.filters.json
+                    - pipeline/clean/opuscleaner/configs/en-ru/mtdata_ELRC-3075-wikipedia_health-v1.en-ru.filters.json
```
Comment (gregtatum) [inline taskcluster/ci/clean-corpus/kind.yml:39]: Question: What's the strategy here? Are we adding all of the filters for all languages and datasets here?

File: pipeline/clean/opuscleaner/generate_filters.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    # TODO: ideally "other" for "deescape-special-chars" should be replaced to <trg> for supported languages
+    with open("configs/default.filters.json") as f:
+        config_str = f.read()
+        config_str = config_str.replace("<src>", src).replace("<trg>", trg)
+        abs_path_patterns = os.path.abspath("configs/remove_frequent_patterns.txt")
+        config_str = config_str.replace("configs/remove_frequent_patterns.txt", abs_path_patterns)
+        return json.loads(config_str)
+
+
+def main() -> None:
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/generate_filters.py:48]: Suggestion: A test for this would be good to have. It would be nice to run a test exercising the CLI arguments. We don't have the first test written using pytest, so the burden is a bit higher. If you don't want to write a test for it, then I could potentially follow-up and write a test for it after my single process work.

I would imagine we would have some kind of test utilities where we could run actual commands like

```py
test_generate_filters():
     results = run_cli('pipeline/clean/opuscleaner/generate_filters.py --src="en" --trg="ru"...')
     assert ...
````

File: pipeline/clean/opuscleaner/clean-corpus.sh [hunk 9]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/clean-corpus.sh:1]: Thought: This command is testable as well, but I'm not going to request a test here at this time.

Comment (gregtatum) [inline pipeline/clean/opuscleaner/clean-corpus.sh:1]: Question: Is there a reason this is in bash rather than python?

File: pipeline/clean/opuscleaner/clean-corpus.sh [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+echo "### Cleaning ${input_prefix} with filter ${filter_path}"
+paste <(${COMPRESSION_CMD} -dc "${input_prefix}.${SRC}.${ARTIFACT_EXT}") \
+      <(${COMPRESSION_CMD} -dc "${input_prefix}.${TRG}.${ARTIFACT_EXT}") |
+opuscleaner-clean \
+  --parallel ${threads} \
+  --batch-size=50000 \
+  --input=- \
+  "${filter_path}" "${SRC}" "${TRG}" \
+  2> "${output_prefix}.opus_filter.log" |
```
Comment (gregtatum) [inline pipeline/clean/opuscleaner/clean-corpus.sh:44]: Question: I guess this outputs the opus log to the artifacts directory.

File: taskcluster/ci/clean-corpus/kind.yml [hunk 11]
```diff
@@ -58,10 +74,15 @@ tasks:
...
+            # We're going to migrate to OpusCleaner in the long term so we can use a simple 'if' in the command
+            # for now without introducing an extra step and complexity
             command:
                 - bash
                 - -c
-                - $VCS_PATH/pipeline/clean/clean-corpus.sh $MOZ_FETCHES_DIR/{dataset_sanitized} /builds/worker/artifacts/{dataset_sanitized} auto {dataset}
+                - >-
+                    pip install -r $VCS_PATH/pipeline/clean/requirements/clean.txt &&
+                    if [ ${USE_OPUSCLEANER} = "true" ]; then dir="clean/opuscleaner"; else dir="clean"; fi &&
+                    $VCS_PATH/pipeline/${dir}/clean-corpus.sh $MOZ_FETCHES_DIR/{dataset_sanitized} /builds/worker/artifacts/{dataset_sanitized} auto {dataset}
```
Comment (gregtatum) [inline taskcluster/ci/clean-corpus/kind.yml:85]: Suggestion: This will fail the new yamllint rules as beeing too long. It's also not very readable as it is written.

Comment (gregtatum) [inline taskcluster/ci/clean-corpus/kind.yml:85]: Thought: This is another example where the bash script vs python argparse makes it difficult to understand how everything is wired together. With named arguments this would be much easier to understand.

**Review by gregtatum — APPROVED** [review]
Suggestion (docs): Since we're adding new tools to the pipeline, it would be nice to start the practice of documenting them a bit more. It would be nice to get a high-level markdown file documenting what is going on with OpusCleaner and why we use it. It would be nice to explain how we layer the configs, and how to start writing and using your own. For this PR it would be nice to at least start this documentation process, and I think it would be fine to file a follow-up to add more docs from there.

In conclusion: I'm going to mark this as approved on my end as I don't see any reason to block it. If you add substantial changes it would be nice for me to look at it again, so feel free to re-flag me for review.

Again, I use https://conventionalcomments.org/ for my reviews, so things like "Thought" is not something that required action, while "Suggestion" is. I use "Thought" when it's nice to have a conversation or when I'm thinking about the bigger problem.

File: pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt (hunk 3 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/configs/remove_frequent_patterns.txt:1]: This file is required for one of the filters and it is a copy-paste from opus-cleaner. https://github.com/hplt-project/OpusCleaner/blob/main/opuscleaner/filters/remove_frequent_patterns.txt

File: pipeline/clean/opuscleaner/clean-corpus.sh (hunk 10 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/clean-corpus.sh:44]: Yes, sometimes the opuscleaner writes to stderr that we can't see in the artifacts. We should address this issue on a higher level probably.

File: pipeline/clean/opuscleaner/clean-corpus.sh (hunk 9 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/clean-corpus.sh:1]: This is a very simple script that just calls CLI tools and does some IO with them, so I didn't see any reason to write it in Python. Also, we don't have a pattern setup of how to write such things in Python efficiently. It shouldn't be hard, but it's just way faster to write this in bash. As a follow up you can try refactoring this later in Python to compare the complexity and the benefits of such scripts.

File: pipeline/clean/opuscleaner/generate_filters.py (hunk 8 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/generate_filters.py:48]: We'll need to set up this first. I filed an issue https://github.com/mozilla/firefox-translations-training/issues/209

File: taskcluster/ci/clean-corpus/kind.yml (hunk 11 above)
Comment (eu9ene) [reply taskcluster/ci/clean-corpus/kind.yml:85]: Let's merge the linting and then fix all the issues it finds.

**Review by gabrielBusta — APPROVED** [review]

File: pipeline/clean/opuscleaner/generate_filters.py (hunk 5 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/generate_filters.py:12]: I filed a ticket with the follow-up work https://github.com/mozilla/firefox-translations-training/issues/210. I hope we can start working on this soon.

**Comment by eu9ene** [comment]
> Suggestion (docs): Since we're adding new tools to the pipeline, it would be nice to start the practice of documenting them a bit more. It would be nice to get a high-level markdown file documenting what is going on with OpusCleaner and why we use it. It would be nice to explain how we layer the configs, and how to start writing and using your own. For this PR it would be nice to at least start this documentation process, and I think it would be fine to file a follow-up to add more docs from there.
> 
> In conclusion: I'm going to mark this as approved on my end as I don't see any reason to block it. If you add substantial changes it would be nice for me to look at it again, so feel free to re-flag me for review.
> 
> Again, I use https://conventionalcomments.org/ for my reviews, so things like "Thought" is not something that required action, while "Suggestion" is. I use "Thought" when it's nice to have a conversation or when I'm thinking about the bigger problem.

I have refactoring of the docs on my list and filed an [issue](https://github.com/mozilla/firefox-translations-training/issues/208) for this.

Also, this PR was meant as just an initial integration and to hand over the work as we discussed. The follow-up work that will be required to complete this is listed [here](https://github.com/mozilla/firefox-translations-training/issues/210). We can refactor things there and experiment with alternative approaches.

File: pipeline/clean/opuscleaner/configs/default.filters.json (hunk 2 above)
Comment (eu9ene) [reply pipeline/clean/opuscleaner/configs/default.filters.json:1]: I created this filter template using the OpusCleaner UI tool. It should be compatible with the [filters](https://github.com/hplt-project/OpusCleaner/tree/main/opuscleaner/filters). I don't know if they have a schema somewhere to validate it. This would be a good follow-up task to validate the schema before running things.

**Comment by eu9ene** [comment]
I added the docs and comments according to the suggestions and fixed an issue with the custom config. Nothing major, so merging.

File: pyproject.toml (hunk 6 above)
Comment (eu9ene) [reply pyproject.toml:10]: yeah, we have legacy conda envs for snakemake, pip compiled requirements in the step directories for TaskCluster and the poetry file. Poetry is more convenient for running local utils, so let's keep using it.
